src.related/mapgen/man/man1/plotter.1
.\" @(#)plotter.1 - 1.1
.nr LL 5.5i
.TH PLOTTER 1 2/20/89 "USGS/OEMG Systems"
.ad b
.hy 1
.SH NAME
plotter - plotting of metagraphic stream
.SH SYNOPSIS
.LP
.B plotter
[ -diorspPD [args] ] [ file(s) ]
.SH DESCRIPTION
.B Plotter
interprets a metagraphic stream and reformats the data
into a form acceptable by the selected graphics device.

Except for the
.B -p
option
the following command line control parameters can appear in any order:
.TP
.BI \-d " name"
.br
.I name
defines the name of the plotting device to which the output
is to be directed.
If
.I name
is omitted,
.B plotter
acquires the name of the device from the
.B GTERM
entry of the processes' environment or the
.B TERM
entry if
.B GTERM is not specified.
If
.I name
or the default terminal is not an implemented device,
.B plotter
proceeds with a dummy device and effectively acts as a no_op.
.TP
.BI \-s " scale"
.br
.I Scale
must be a floating point value which will scale the coordinates
values of the input metagraphic stream.
.TP
.BI \-o " output"
.br
This option directs the device dependent graphic command to file
.I output.
This option is only required for non-terminal devices not
defined in the users environment setting for GRAPHB.
In all cases, it will redirect the plotter command stream to
the named file.
.TP
.BI \-i " m.n"
.br
This option only applies to the interactive use of
.B plotter
by programs performing graphics.
It should not be employed by
.I shell
execution of
.B plotter.
.I M.n
are automatically generated by graphics library software.
.I M
is the file descriptor for the input metagraphic data pipe to
.B plotter
and
.I n
is the file descriptor of the return data pipe to the calling
program.
.TP
.BI \-q
.br
This option alerts interactive display drivers not to sound
console bell and wait for a <CR> at the end of a plot.
Its primary purpose is with cursor control applications which
hold the screen with cursor input and where plot termination
is at user's cursor input control.
.TP
.BI \-r
.br
If this option is selected, the x and y axis are reversed on
the plotting device.
.TP
.BI \-[p|P] "m:n[,m:n]"
.br
Mapping the mechanical pens selected in the original
overlay files to new mechanical pens may be accomplished with
this option.
This option is sensitive to its position on the run line
and only affects overlay files that follow its specification.
It may be employed on the run-line as often as meaningful.
.IP
The argument pair
.I m:n
(immediately following the
.B -p
or
.B -P
)
consist of the original mechanical pen number
.I m
and the new mechanical pen number
.I n
to be employed in this plot.
.I M
may also denote a range of original mechanical pen numbers when a -
is employed.
For example: 3-6:0 maps original values 3 through 6 to new pen 0,
0-3:5 or -3:5 maps pens 0 through 3 to new pen 5, and
3-:2 maps pen 3 through 255 to pen 2.
.IP
The
.B -P
option applies to interactive applications where the pen mapping
is to be performed once before any graphic processing.
.TP
.BI \-Dstring
.br
.I String
is information passed to the selected device driver.
Reference to specific driver documentation must be made for details
of content.
.TP
.BI file(s)
.br
The files named must contain metagraphic commands compatible
with this system (see Device Independent Graphics manual).
The files are processed in a left to right order.
A \- may be employed to designate input from
.I stdin.
If no files are given and option
.B \-i
is not employed,
.I stdin
is assumed as the source of the metagraphic stream.
.SH ENVIRONMENT
.P
.B Plotter
requires the environment entry
.B GRAPHB
which is also employed by application software employing this
graphics system.
The general
.B csh
method of initializing this environment parameter is:
.RS 4
.sp
\f(CWsetenv GRAPHB prog_path:fonts_path:def_font[:dev,file]:\fR
.sp
.RE
where
.I prog_path
is the full path name of program
.B plotter,
.I fonts_path
is the directory containing the font definition data, and
.I def_font
is the default font name in
.I fonts_path
employed when only a - font name is used.
.I Dev,file
is an optional list of plotting devices and their output file name.
Note that each plotter entry must be separated by a : and
associated file delimited with a ,.
The following is an example entry:
.RS 4
.sp
\f(CWsetenv GRAPHB /graph/plotter:/graph/fonts/:sr:hp7475,/dev/ttym2:\fR
.RE
.SH EXAMPLE
plotter file1 file2 -s 4
.sp
will generate a composite plot of both files on the user's terminal
and will have the coordinates scaled by a factor of 4.
Obviously, the terminal should be a device capable of graphic
output.
.sp
\f(CWplotter file1 -p1:0 file2 -sdo .25 calcomp caltemp\fR
.sp
will scale the meta-graphic files by 1/4 and output the 
.I calcomp
plotter control to the disc file
.I caltemp.
The mechanical pen 1 of metagraphic
.I file2
is mapped to pen 0;
.SH SEE ALSO
 Device Independent Vector Graphics manual.
 Documentation of specific graphic devices.
 GRAPHICS(3)
.SH DIAGNOSTICS
If an invalid graphics device is selected, a message is
output to
.I stderr
and all input data are ignored.
Additional error conditions are available only though
bidirectional linkage with controlling process.
.SH AUTHOR/MAINTENANCE
Gerald I. Evenden, USGS, Woods Hole, MA 02543


CHANGES
- GISLIB 
  - G_warning(), G_fatal_error() - send mail only if eviroment variable
                                   GRASS_ERROR_MAIL is set
  - G_debug() - new, print debugging info on selected level
  - G_plot_icon() - new, plot selected mark
- DBMI 
  - dbmscap file is not used, drivers are searched in dist.$ARCH/driver/db
  - new dbf driver
- vector library completely rewritten (multilayer, multiformat, 3D)
- new make system


src.related/mapgen/src/graphics/plotter/plotter.c
#ifndef lint
static char *SCCSID = "@(#)plotter.c	OEMG v.2.1";
#endif
#include <signal.h>
#include <varargs.h>
#define PLOTTER
#define MAIN_PROG
#include "graphics.h"
#include "grerror.h"
#include "plotter.h"
extern void exit();
extern long strtol();
extern char *strcpy();
	extern struct DEV_LIST
dev_list[];
	BASE
base = { 0, 0, NULL, };
	long
base_x, base_y;
	char
*font_dir[] = {"\0", "\0"};
	static char
*env_name = _GENVB,
*env_base,
*initmap,
*ofile;	/* output file name */

PEN *pen = 0;	/* pen control block pointer */

int error = 0;	/* error status */
	unsigned
pmaplist[] = { /* mechanical pen mapping array */
	  0,  1,  2,  3,  4,  5,  6,  7,  8,  9,
	 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
	 20, 21, 22, 23, 24, 25, 26, 27, 28, 29,
	 30, 31, 32, 33, 34, 35, 36, 37, 38, 39,
	 40, 41, 42, 43, 44, 45, 46, 47, 48, 49,
	 50, 51, 52, 53, 54, 55, 56, 57, 58, 59,
	 60, 61, 62, 63, 64, 65, 66, 67, 68, 69,
	 70, 71, 72, 73, 74, 75, 76, 77, 78, 79,
	 80, 81, 82, 83, 84, 85, 86, 87, 88, 89,
	 90, 91, 92, 93, 94, 95, 96, 97, 98, 99,
	100,101,102,103,104,105,106,107,108,109,
	110,111,112,113,114,115,116,117,118,119,
	120,121,122,123,124,125,126,127,128,129,
	130,131,132,133,134,135,136,137,138,139,
	140,141,142,143,144,145,146,147,148,149,
	150,151,152,153,154,155,156,157,158,159,
	160,161,162,163,164,165,166,167,168,169,
	170,171,172,173,174,175,176,177,178,179,
	180,181,182,183,184,185,186,187,188,189,
	190,191,192,193,194,195,196,197,198,199,
	200,201,202,203,204,205,206,207,208,209,
	210,211,212,213,214,215,216,217,218,219,
	220,221,222,223,224,225,226,227,228,229,
	230,231,232,233,234,235,236,237,238,239,
	240,241,242,243,244,245,246,247,248,249,
	250,251,252,253,254,255
};

XY *(*device)();
	static void
onintr() { /* interrupt trap */
	if (device) (*device)(D_PANIC);
	exit (1);
}
	static void
bomb(va_alist) va_dcl {
	va_list vargs;
	int opt;
	extern void perror();

	va_start(vargs);
	opt = va_arg(vargs, int);
	(void)fputs("plotter failure, Version:",stderr);
	(void)fputs(VERSION,stderr);
	if (opt & PERR) perror(va_arg(vargs, char *));
	opt &= ~PERR;
	while (opt-- > 0)
		(void)fputs(va_arg(vargs, char *), stderr);
	(void)fputc('\n',stderr);
	va_end(vargs);
	onintr();
}
	static char * /* word splitter for env data */
nextenv(code) {
	char *s, *strchr();

	if (env_base) {
		s = env_base;
		if (env_base = strchr(env_base, code))
			*env_base++ = '\0';
	} else
		s = "\0";
	return s;
}
	static void
setdev(name) char *name; {
	int i;
	char *getenv(), *s;
		/* if name not given, use terminal's */
	if (! name)
		if (!(name = getenv("GTERM")))
			name = getenv("TERM");
		/* look up name in list */
	for (i = 0; s = dev_list[i].name ; i++)
		if (!strcmp(s, name)) {
			device = dev_list[i].dev;
			Dglobal.model_no = dev_list[i].model;
			break;
		}
	/* couldn't find a legit name. use last in list! */
	if (! s)
		device = dev_list[i-1].dev;
	else if (!ofile) /* hard on-line device */
		while (*(s = nextenv(','))) {
			if (!strcmp(dev_list[i].name, s)) {
				ofile = nextenv(':');
				break;
			}
			(void)nextenv(':');
		}
	if (ofile && !freopen(ofile,"w+",stdout))
		bomb(PERR+2, "output file", "\nrunline -o ",ofile);
}
main(argc, argv) int argc; char *argv[]; {
	double atof();
	long strtol();
	XY *p;
	int i;
	char *pstr, *devname = (char *)0;
	char **aargv, *arg, *getenv(), *strchr();
	int aargc;

	/* set interupt trap */
	if (signal(SIGINT, SIG_IGN) != SIG_IGN)
		(void)signal(SIGINT, onintr);
	if (!(env_base = getenv(env_name)))
		bomb(3,"environ: ", env_name, " not found");
	/* set default font directory and font pointer */
	(void)nextenv(':'); /* skip over program's name */
	init_pens();
	for (i = 0; i < 2; ++i)
		font_dir[i] = nextenv(':');
	aargv = argv;
	aargc = 0;
	while (--argc > 0) { /* process options */
		if (**++argv == '-') for (arg = *argv;;) {
			switch (*++arg) {
			case '\0': /* stdin file */
				if (arg[-1] == '-')
					aargv[aargc++] = "-";
				break;
			case 'o': /* output file */
				if (--argc > 0)
					ofile = *++argv;
				continue;
			case 'X': /* x base offset */
				if (--argc > 0)
					base_x = strtol(*++argv,0,0);
				continue;
			case 'Y': /* y base offset */
				if (--argc > 0)
					base_y = strtol(*++argv,0,0);
				continue;
			case 's': /* scale factor */
				if (--argc > 0)
					Dglobal.scale = atof(*++argv);
				continue;
			case 'd': /* device name */
				if (--argc > 0)
					devname = *++argv;
				continue;
			case 'D': /* device arg */
				Dglobal.dargv[Dglobal.dargc++] = ++arg;
				break;
			case 'r': /* reverse axis */
				Dglobal.reverse = 1;
				continue;
			case 'q': /* mute <bell prompt> */
				Dglobal.quiet = 1;
				continue;
			case 'i': /* interactive */
				if (--argc > 0) {
					pstr = "-               ";
					(void)strcpy(pstr+1,*++argv);
					aargv[aargc++] = pstr;
/*DEBUG fprintf(stderr,"plotter %s\n",pstr);*/
					continue;
				} else
					bomb(1,"no interactive link args\n");
				continue;
			case 'p': /* pen mapping */
				aargv[aargc++] = *argv;
				break;
			case 'P': /* initial pen mapping */
				initmap = *argv + 2;
				break;
			default: /* invalid argument */
				bomb(2, "invalid arg: ", *argv);
			}
			break;
		} else /* input file */
			aargv[aargc++] = *argv;
	}
	if (initmap) pmapin(initmap);
	if (! aargc ) aargv[aargc++] = "-";
	setdev(devname);
	if ((p = (*device)(D_INIT)) == NULL)
		bomb(1,"device initialization failure");
	windinit((long)(p->x), (long)(p->y));
	while (aargc--) {
		if (aargv[0][0] == '-' && aargv[0][1] == 'p')
			pmapin(*(aargv++)+2);
		else {
			initincl(*aargv++);
			pltparse();
		}
	}
	(*device)(D_DONE);
	exit(0);
#ifdef lint
return 1;
#endif
}
#define MAXINCLUDE 5
#define MAXSTRING  256

static file_no, direct;
static FILE *ilist[MAXINCLUDE], *answr;
static long xb[MAXINCLUDE], yb[MAXINCLUDE];

/* initialize include pre-processor */
	static
initincl(file) char *file; {
	char *strchr();

	if (direct = (*file == '-' && file[1] != '\0')) {
		if ((ilist[0] = fdopen(atoi(++file),"r")) == NULL)
		      bomb(PERR, "pipe input open failure");
		file = strchr(file, '.');
		if (!file || (answr = fdopen(atoi(++file),"w")) == NULL)
		      bomb(PERR, "pipe output open failure\n");
	}
	else if (*file == '-')
		ilist[0] = stdin;
	else if ((ilist[0] = fopen(file,"r")) == NULL)
		bomb(PERR+2, "input file","\nrunline name:",file);
	file_no = 0;
	base.x = base_x;
	base.y = base_y;
}
metain() { /* get next meta-graphic byte */
	static c;

	while (file_no >= 0 && (c = fgetc(ilist[file_no])) == EOF) {
		base.x = xb[file_no];
		base.y = yb[file_no];
		(void)fclose(ilist[file_no--]);
	}
	return(c);
}
metacmd() { /* get next meta-graphic and interpret as possible
	include command byte */
	int c;
	char *name, *mstring();

	while ((c = metain()) == INCL) {
		name = mstring();
		if (++file_no < MAXINCLUDE) {
			if ((ilist[file_no] = fopen(name, "r")) == NULL) {
				error = E_INCL;
				file_no--;
			} else {
				xb[file_no] = base.x;
				yb[file_no] = base.y;
				base.x = base_x;
				base.y = base_y;
			}
		} else {
			error = E_INCLOV;
			file_no--;
		}
	}
	return(c);
}
static char str[MAXSTRING+1];
	char *
mstring() {
	register c, i;
	register char *s;

	s = str;
	for ( i = 0; (c = metain()) != '\0' && c != EOF ; )
		if (i < MAXSTRING) *s++ = c;
	*s = '\0';
	return(str);
}
	FILE * /* return file pointer to applications program */
ufile() { return ( (!file_no && direct) ? answr : NULL); }
	static /* decode pen mapping */
pmapin(s) char *s; {
	int low, high, new;

	do {
		low = *s == '-' ? 0 : strtol(s, &s, 10);
		if (*s == '-') {
			++s;
			high = *s == ':' ? 255 : strtol(s, &s, 10);
		} else
			high = low;
		if (*s == ':') {
			++s;
			new = strtol(s, &s, 10);
		} else
			bomb(1,"improper -p list\n");
		for ( ; low <= high ; ++low)
			pmaplist[low] = new;
	} while (*s++ == ',');
}
